### PR TITLE
lightning: add add-index-by-sql

### PR DESCRIPTION
**Pull request by gozssky**
### What is changed, added or deleted? (Required)

Add add-index-by-sql.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v7.0 (TiDB 7.0 versions)
- [ ] v6.6 (TiDB 6.6 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.4 (TiDB 6.4 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: tidb-lightning/tidb-lightning-configuration.md [hunk 1]
```diff
@@ -146,6 +146,10 @@ addr = "172.16.31.10:8287"
 # 默认值为 MaxInt64 字节，即 9223372036854775807 字节。
 # disk-quota = "10GB" 
 
+# Physical Import Mode 是否通过 SQL 方式添加索引。默认根据 TiDB 版本自动选择，如果 TiDB 版本 >= v7.0.0, 则 TiDB Lightning 会使用 SQL 方式添加索引。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-configuration.md:149]: 早于 v7.0.0 版本的时候默认是 false 吗？

Comment (gozssky) [reply tidb-lightning/tidb-lightning-configuration.md:149]: 是的

Comment (hfxsd) [inline tidb-lightning/tidb-lightning-configuration.md:149]:
```suggestion
# Physical Import Mode 是否通过 SQL 方式添加索引。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false`。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会使用 SQL 方式添加索引。
```

Comment (hfxsd) [reply tidb-lightning/tidb-lightning-configuration.md:149]: OK，我更新了描述，PTAL。

File: tidb-lightning/tidb-lightning-physical-import-mode-usage.md [hunk 2]
```diff
@@ -40,6 +40,10 @@ sorted-kv-dir = "./some-dir"
 # 限制 TiDB Lightning 向每个 TiKV 节点写入的带宽大小，默认为 0，表示不限制。
 # store-write-bwlimit = "128MiB"
 
+# 配置是否通过 SQL 方式添加索引。默认根据 TiDB 版本自动选择，如果 TiDB 版本 >= v7.0.0, 则 TiDB Lightning 会使用 SQL 方式添加索引。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-physical-import-mode-usage.md:43]:
```suggestion
# 配置是否通过 SQL 方式添加索引。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false`。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会使用 SQL 方式添加索引。
```

**Review by hfxsd — APPROVED** [review]

File: tidb-lightning/tidb-lightning-configuration.md [hunk 3]
```diff
@@ -146,6 +146,10 @@ addr = "172.16.31.10:8287"
 # 默认值为 MaxInt64 字节，即 9223372036854775807 字节。
 # disk-quota = "10GB" 
 
+# Physical Import Mode 是否通过 SQL 方式添加索引。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false`。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会使用 SQL 方式添加索引。
+# 通过 SQL 方式添加索引的优点是可以快速导入数据，即使索引添加失败，也不会影响数据的一致性。
+#add-index-by-sql = true
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-configuration.md:151]:
```suggestion
# add-index-by-sql = true
```

File: tidb-lightning/tidb-lightning-physical-import-mode-usage.md [hunk 4]
```diff
@@ -40,6 +40,10 @@ sorted-kv-dir = "./some-dir"
 # 限制 TiDB Lightning 向每个 TiKV 节点写入的带宽大小，默认为 0，表示不限制。
 # store-write-bwlimit = "128MiB"
 
+# 配置是否通过 SQL 方式添加索引。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false`。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会使用 SQL 方式添加索引。
+# 通过 SQL 方式添加索引的优点是可以快速导入数据，即使索引添加失败，也不会影响数据的一致性。
+#add-index-by-sql = true
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-physical-import-mode-usage.md:45]:
```suggestion
# add-index-by-sql = true
```

File: tidb-lightning/tidb-lightning-physical-import-mode.md [hunk 5]
```diff
@@ -18,15 +18,17 @@ Physical Import Mode 对应的后端模式为 `local`。
...
+    当 TiDB 集群版本 >= v7.0.0，`tidb-lightning` 默认使用 SQL 接口添加索引，并且会在导入数据前移除目标表的所有次级索引。
+
 3. 每张表都会被分割为多个连续的**区块**，这样来自大表 (200 GB+) 的数据就可以多个并发导入。
 
 4. `tidb-lightning` 会为每一个区块准备一个“引擎文件 (engine file)”来处理键值对。`tidb-lightning` 会并发读取 SQL dump，将数据源转换成与 TiDB 相同编码的键值对，然后将这些键值对排序写入本地临时存储文件中。
 
 5. 当一个引擎文件数据写入完毕时，`tidb-lightning` 便开始对目标 TiKV 集群数据进行分裂和调度，然后导入数据到 TiKV 集群。
 
-    引擎文件包含两种：**数据引擎**与**索引引擎**，各自又对应两种键值对：行数据和次级索引。通常行数据在数据源里是完全有序的，而次级索引是无序的。因此，数据引擎文件在对应区块写入完成后会被立即上传，而所有的索引引擎文件只有在整张表所有区块编码完成后才会执行导入。
+    引擎文件包含两种：**数据引擎**与**索引引擎**，各自又对应两种键值对：行数据和次级索引。通常行数据在数据源里是完全有序的，而次级索引是无序的。因此，数据引擎文件在对应区块写入完成后会被立即上传，而所有的索引引擎文件只有在整张表所有区块编码完成后才会执行导入。 注意当 `tidb-lightning` 使用 SQL 接口添加索引时，索引引擎将不会写入数据，因为此时目标表的次级索引已经在第 2 步中被移除。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-physical-import-mode.md:29]:
```suggestion
    引擎文件包含两种：**数据引擎**与**索引引擎**，各自又对应两种键值对：行数据和次级索引。通常行数据在数据源里是完全有序的，而次级索引是无序的。因此，数据引擎文件在对应区块写入完成后会被立即上传，而所有的索引引擎文件只有在整张表所有区块编码完成后才会执行导入。 
    
    注意当 `tidb-lightning` 使用 SQL 接口添加索引时，索引引擎将不会写入数据，因为此时目标表的次级索引已经在第 2 步中被移除。
```

File: tidb-lightning/tidb-lightning-physical-import-mode-usage.md [hunk 6]
```diff
@@ -40,6 +40,10 @@ sorted-kv-dir = "./some-dir"
 # 限制 TiDB Lightning 向每个 TiKV 节点写入的带宽大小，默认为 0，表示不限制。
 # store-write-bwlimit = "128MiB"
 
+# 配置是否通过 SQL 方式添加索引。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false`。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会使用 SQL 方式添加索引。
```
[the remark this answers is not in the document]
Comment (Frank945946) [reply tidb-lightning/tidb-lightning-physical-import-mode-usage.md:43]: @gozssky  PTAL

File: tidb-lightning/tidb-lightning-physical-import-mode-usage.md [hunk 7]
```diff
@@ -40,6 +40,10 @@ sorted-kv-dir = "./some-dir"
...
+# 通过 SQL 方式添加索引的优点是可以快速导入数据，即使索引添加失败，也不会影响数据的一致性。
```
[the remark this answers is not in the document]
Comment (Frank945946) [reply tidb-lightning/tidb-lightning-physical-import-mode-usage.md:44]: @gozssky PTAL

**Comment by gozssky** [comment]
@hfxsd There are new changes. PTAL again.

File: tidb-lightning/tidb-lightning-configuration.md [hunk 8]
```diff
@@ -146,6 +146,10 @@ addr = "172.16.31.10:8287"
 # 默认值为 MaxInt64 字节，即 9223372036854775807 字节。
 # disk-quota = "10GB" 
 
+# Physical Import Mode 是否通过 SQL 方式添加索引。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false`，表示 Lightning 会将行数据以及索引数据都编码成 kv paris 后一同导入 TiKV，和历史版本保持一致。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会在导入数据完成后，使用 add index 的 SQL 来添加索引。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-configuration.md:149]:
```suggestion
# Physical Import Mode 是否通过 SQL 方式添加索引。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false`，表示 TiDB Lightning 会将行数据以及索引数据都编码成 KV pairs 后一同导入 TiKV，和历史版本保持一致。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会在导入数据完成后，使用 add index 的 SQL 来添加索引。
```

File: tidb-lightning/tidb-lightning-physical-import-mode-usage.md [hunk 9]
```diff
@@ -40,6 +40,10 @@ sorted-kv-dir = "./some-dir"
 # 限制 TiDB Lightning 向每个 TiKV 节点写入的带宽大小，默认为 0，表示不限制。
 # store-write-bwlimit = "128MiB"
 
+# Physical Import Mode 为目标表添加索引的方式。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false` 表示仍然会用 Lightning 将所需导入的数据以及索引编码成 kv pairs 后再一同导入到 TiKV。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会导完数据后，再使用 add index 的 SQL 来添加索引。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-physical-import-mode-usage.md:43]:
```suggestion
# Physical Import Mode 为目标表添加索引的方式。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false` 表示仍然会用 TiDB Lightning 将所需导入的数据以及索引编码成 KV pairs 后再一同导入到 TiKV。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会导完数据后，再使用 add index 的 SQL 来添加索引。
```

File: tidb-lightning/tidb-lightning-configuration.md [hunk 10]
```diff
@@ -146,6 +146,10 @@ addr = "172.16.31.10:8287"
 # 默认值为 MaxInt64 字节，即 9223372036854775807 字节。
 # disk-quota = "10GB" 
 
+# Physical Import Mode 是否通过 SQL 方式添加索引。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false`，表示 TiDB Lightning 会将行数据以及索引数据都编码成 KV pairs 后一同导入 TiKV，和历史版本保持一致。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会在导入数据完成后，使用 add index 的 SQL 来添加索引。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-configuration.md:149]:
```suggestion
# Physical Import Mode 是否通过 SQL 方式添加索引。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false`，表示 TiDB Lightning 会将行数据以及索引数据都编码成 KV pairs 后一同导入 TiKV，实现机制和历史版本保持一致。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会在导入数据完成后，使用 add index 的 SQL 来添加索引。
```

File: tidb-lightning/tidb-lightning-physical-import-mode-usage.md [hunk 11]
```diff
@@ -40,6 +40,10 @@ sorted-kv-dir = "./some-dir"
 # 限制 TiDB Lightning 向每个 TiKV 节点写入的带宽大小，默认为 0，表示不限制。
 # store-write-bwlimit = "128MiB"
 
+# Physical Import Mode 为目标表添加索引的方式。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false` 表示仍然会用 TiDB Lightning 将所需导入的数据以及索引编码成 KV pairs 后再一同导入到 TiKV。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会导完数据后，再使用 add index 的 SQL 来添加索引。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-physical-import-mode-usage.md:43]:
```suggestion
# Physical Import Mode 为目标表添加索引的方式。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false` 表示仍然会用 TiDB Lightning 将所需导入的数据以及索引编码成 KV pairs 后再一同导入到 TiKV，实现机制和历史版本保持一致。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会导完数据后，再使用 add index 的 SQL 来添加索引。
```

File: tidb-lightning/tidb-lightning-physical-import-mode.md [hunk 12]
```diff
@@ -18,15 +18,19 @@ Physical Import Mode 对应的后端模式为 `local`。
...
 
 5. 当一个引擎文件数据写入完毕时，`tidb-lightning` 便开始对目标 TiKV 集群数据进行分裂和调度，然后导入数据到 TiKV 集群。
 
-    引擎文件包含两种：**数据引擎**与**索引引擎**，各自又对应两种键值对：行数据和次级索引。通常行数据在数据源里是完全有序的，而次级索引是无序的。因此，数据引擎文件在对应区块写入完成后会被立即上传，而所有的索引引擎文件只有在整张表所有区块编码完成后才会执行导入。
+    引擎文件包含两种：**数据引擎**与**索引引擎**，各自又对应两种键值对：行数据和次级索引。通常行数据在数据源里是完全有序的，而次级索引是无序的。因此，数据引擎文件在对应区块写入完成后会被立即上传，而所有的索引引擎文件只有在整张表所有区块编码完成后才会执行导入。 
+    
+    注意当 `tidb-lightning` 使用 SQL 接口添加索引时，索引引擎将不会写入数据，因为此时目标表的次级索引已经在第 2 步中被移除。
 
-6. 整张表相关联的所有引擎文件完成导入后，`tidb-lightning` 会对比本地数据源及下游集群的校验和 (checksum)，确保导入的数据无损，然后让 TiDB 分析 (`ANALYZE`) 这些新增的数据，以优化日后的操作。同时，`tidb-lightning` 调整 `AUTO_INCREMENT` 值防止之后新增数据时发生冲突。
+6. 整张表相关联的所有引擎文件完成导入后，`tidb-lightning` 会对比本地数据源及下游集群的校验和 (checksum)，确保导入的数据无损，然后添加缺失的次级索引或让 TiDB 分析 (`ANALYZE`) 这些新增的数据，以优化日后的操作。同时，`tidb-lightning` 调整 `AUTO_INCREMENT` 值防止之后新增数据时发生冲突。
```
Comment (Frank945946) [inline tidb-lightning/tidb-lightning-physical-import-mode.md:33]:
```suggestion
6. 整张表相关联的所有引擎文件完成导入后，`tidb-lightning` 会对比本地数据源及下游集群的校验和 (checksum)，确保导入的数据无损，然后添加在第 2 步中被移除的次级索引或让 TiDB 分析 (`ANALYZE`) 这些新增的数据，以优化日后的操作。同时，`tidb-lightning` 调整 `AUTO_INCREMENT` 值防止之后新增数据时发生冲突。
```

**Review by Frank945946 — APPROVED** [review]

File: tidb-lightning/tidb-lightning-physical-import-mode-usage.md [hunk 13]
```diff
@@ -40,6 +40,10 @@ sorted-kv-dir = "./some-dir"
 # 限制 TiDB Lightning 向每个 TiKV 节点写入的带宽大小，默认为 0，表示不限制。
 # store-write-bwlimit = "128MiB"
 
+# Physical Import Mode 为目标表添加索引的方式。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false` 表示仍然会用 TiDB Lightning 将所需导入的数据以及索引编码成 KV pairs 后再一同导入到 TiKV，实现机制和历史版本保持一致。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会导完数据后，再使用 add index 的 SQL 来添加索引。
```
Comment (qiancai) [inline tidb-lightning/tidb-lightning-physical-import-mode-usage.md:43]:
```suggestion
# Physical Import Mode 是否通过 SQL 方式添加索引。默认根据 TiDB 版本自动选择。如果 TiDB 版本早于 v7.0.0，则默认为 `false` 表示仍然会用 TiDB Lightning 将所需导入的数据以及索引编码成 KV pairs 后再一同导入到 TiKV，实现机制和历史版本保持一致。从 v7.0.0 版本开始，默认为 `true`，即 TiDB Lightning 会导完数据后，再使用 add index 的 SQL 来添加索引。
```

Comment (qiancai) [reply tidb-lightning/tidb-lightning-physical-import-mode-usage.md:43]: 和 tidb-lightning/tidb-lightning-configuration.md 中的描述保持一致

**Review by qiancai — APPROVED** [review]
other LGTM

**Pull request merged**